tool/tctl/common/plugin/awsic.go

```diff
@@ -0,0 +1,242 @@
+	"github.com/gravitational/teleport/lib/utils"
+	awsutils "github.com/gravitational/teleport/lib/utils/aws"
+	toolcommon "github.com/gravitational/teleport/tool/common"
+)
+
+type awsICArgs struct {
+	cmd                  *kingpin.CmdClause
+	defaultOwners        []string
+	scimToken            string
+	scimURL              *url.URL
+	region               string
+	arn                  string
+	useSystemCredentials bool
+	userOrigins          []string
+	userLabels           toolcommon.LabelFilters
+	groupNameFilters     []string
+	accountNameFilters   []string
+	accountIDFilters     []string
+}
+
+func (a *awsICArgs) validate() error {
+	if !awsutils.IsKnownRegion(a.region) {
+		return trace.BadParameter("unknown AWS region: %s", a.region)
+	}
+
+	if a.scimToken == "" {
+		return trace.BadParameter("SCIM token must not be empty")
+	}
+
+	if !a.useSystemCredentials {
+		return trace.BadParameter("only AWS Local system credentials are supported")
+	}
+
+	return nil
+}
+
+// parseAWSICNameFilters validates that all elements of the supplied [names] slice
+// are valid regexes or globs and wraps them in [types.AWSICResourceFilter]s for
+// inclusion in a [types.PluginAWSICSettings].
+//
+// We are using a manual validator here rather than the canonical one defined
+// in the AWS IC integration itself, because those filter tools are not
+// available to OSS builds of tctl.
+//
+// TODO(tcsc): Move the filter validation somewhere that the that the OSS build can access it.
+func parseAWSICNameFilters(names []string) ([]*types.AWSICResourceFilter, error) {
+	var filters []*types.AWSICResourceFilter
+	for _, n := range names {
+		if _, err := utils.CompileExpression(n); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		filters = append(filters, &types.AWSICResourceFilter{
+			Include: &types.AWSICResourceFilter_NameRegex{NameRegex: n},
+		})
+	}
+	return filters, nil
+}
+
+func (a *awsICArgs) parseGroupFilters() ([]*types.AWSICResourceFilter, error) {
+	return parseAWSICNameFilters(a.groupNameFilters)
+}
+
+func (a *awsICArgs) parseAccountFilters() ([]*types.AWSICResourceFilter, error) {
+	filters, err := parseAWSICNameFilters(a.accountNameFilters)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	for _, id := range a.accountIDFilters {
+		filters = append(filters, &types.AWSICResourceFilter{
+			Include: &types.AWSICResourceFilter_Id{Id: id},
+		})
+	}
+
+	return filters, nil
+}
+
+func (a *awsICArgs) parseUserFilters() ([]*types.AWSICUserSyncFilter, error) {
+	var result []*types.AWSICUserSyncFilter
+
+	if len(a.userOrigins) > 0 {
+		result = make([]*types.AWSICUserSyncFilter, 0, len(a.userOrigins))
+		for _, origin := range a.userOrigins {
+			result = append(result, &types.AWSICUserSyncFilter{
+				Labels: map[string]string{types.OriginLabel: origin},
+			})
+		}
+	}
+
+	if len(a.userLabels) > 0 {
+		result = append(make([]*types.AWSICUserSyncFilter, 0, len(result)+len(a.userLabels)), result...)
```

Comment:
I think this is correct but it took me 4 or 5 reads to really understand what was happening here. Is there a simpler way to write this?

Reply:
Turns out that `slices.Grow()` exists.